File: external/cnem_03-10-17/python/tuto/tuto_meca2d.py
##
## Copyright (C) 2003-2011
## Lounes ILLOUL (illoul_lounes@yahoo.fr)
## Philippe LORONG (philippe.lorong@ensam.eu)
## Arts et Metiers ParisTech, Paris, France
##
## $Revision$ - $Date$

import sys
import os
import re
import math
import cPickle
import numpy as np
import scipy as sp
from scipy.sparse import coo_matrix
from scipy.sparse.linalg.dsolve import spsolve
from scipy.sparse.linalg.isolve import gmres

sys.path.append('../bin')

import load_node_front
import py_cnem2d
import check_gs
import py_assembling
import outvar

#-------------------------------------------------------------------------------
# read nodes and boundary :
print '\n--------------------------------------------------------------------------------\n'
print 'read node and boundary...\n'

[XY_Noeud,Nb_Noeud_Front,Ind_Noeud_Front]=load_node_front.load2d('data2d_big/')

#-------------------------------------------------------------------------------
# call scni cnem2d :
print '\n--------------------------------------------------------------------------------\n'

[GS,XY_Noeud,Ind_Noeud_Front,Tri,INNO,INON]=py_cnem2d.scni(XY_Noeud,Nb_Noeud_Front,Ind_Noeud_Front)

#print '\ncheck scni...\n'
#check_gs.check(GS,XY_Noeud)

#-------------------------------------------------------------------------------
#assembling for mechanic :
print '\n--------------------------------------------------------------------------------\n'
print 'assembling\n'

n=len(GS[0])

E_NU=list()
E_NU.append((200.e9,0.266))

[B,C,V]=py_assembling.meca(GS,E_NU)
MK=B.transpose()*(V*C)*B

#-------------------------------------------------------------------------------
#solve case :
print '\n--------------------------------------------------------------------------------\n'
print 'solve case...\n'

#boundary condition definition :
#   U,V deplacement = 0 on Y=0   
#   U = 0 and V=0.1 on Y=1 

v=0.1

n_y_0=list()
n_y_1=list()
n_c_ext=Ind_Noeud_Front[0:Nb_Noeud_Front[0]]

for i in n_c_ext:
    if abs(XY_Noeud[i,1])<1.e-6:
        n_y_0.append(i)
    elif abs(XY_Noeud[i,1]-1.)<1.e-6:
        n_y_1.append(i)

n_y_0=np.array(n_y_0)
n_y_1=np.array(n_y_1)

DDL_U_y_0=n_y_0*2
DDL_V_y_0=n_y_0*2+1
DDL_U_y_1=n_y_1*2
DDL_V_y_1=n_y_1*2+1

UV=np.zeros(2*n)

UV[DDL_U_y_0]=0.
UV[DDL_U_y_0]=v
UV[DDL_V_y_1]=0.
UV[DDL_V_y_1]=0.

DDL_bloc=np.hstack([DDL_U_y_0,DDL_V_y_0,DDL_U_y_1,DDL_V_y_1])
DDL_bloc=np.unique(DDL_bloc)
DDL_Tot=np.array(range(2*n))
DDL_Libre=np.setdiff1d(DDL_Tot,DDL_bloc)

b=-(MK[:,DDL_bloc]*UV[DDL_bloc])
MK=MK[DDL_Libre,:]
MK=MK[:,DDL_Libre]

#solve the linear problem

UV[DDL_Libre]=spsolve(MK,b[DDL_Libre])
#UV[DDL_Libre]=gmres(MK,b[DDL_Libre])[0]

Strain=B*UV 
Stress=C*Strain;
Stress=Stress.reshape((np.size(Stress,0)/3,3))
UV=UV.reshape((np.size(UV,0)/2,2))

#-------------------------------------------------------------------------------
# write the result :
print '\n--------------------------------------------------------------------------------\n'
print 'write result to vtk format for paraview'

outvar.vtk(XY_Noeud,Tri,Stress,UV,0)

#-------------------------------------------------------------------------------

File: external/cnem_03-10-17/matlab/tuto2d/tuto_meca2d.m
% Copyright (C) 2003-2011
% Lounes ILLOUL (illoul_lounes@yahoo.fr)
% Philippe LORONG (philippe.lorong@ensam.eu)
% Arts et Metiers ParisTech, Paris, France

clear all;
close all; 
clc;
addpath('../bin');

%% load domain (nodes + boundary(s)) and  draw boundary(s)

load('data_small');% the domain : nodes + bounary(s) 

% plot 
% boundarys

figure;
hold on;
axis equal;

plot_boundary(XY_Node,Nb_Node_Front,Ind_Node_Front);

% node

plot(XY_Node(:,1),XY_Node(:,2),'.','color','green');

hold off;

%% SCNI integration

[GS,XY_Node,Ind_Node_Front,Ind_Node_Tri,Ind_Node_New_Old,Ind_Node_Old_New,Cel]=...
m_cnem2d_scni(XY_Node,Nb_Node_Front,Ind_Node_Front);

%% mechanical problem : 

%% Assembling

n=size(XY_Node,1);

E_NU=[200e9,0.266];

% e_nu_left=[200e9,0.266];
% e_nu_right=[20e9,0.266];
% 
% E_NU=zeros(n,2);
% for i=1:n
%     if XY_Node(i,1)<0.5
%         E_NU(i,:)=e_nu_left;
%     else
%         E_NU(i,:)=e_nu_right;
%     end
% end

[B,C,V]= m_assembling_meca(GS,E_NU);
K_Mat_meca=B'*V*C*B;

%% boundary condition definition

V_bh=0.; % Y displacement at x==0. and x==1.
U_b=0.; % X Displacement at x==0.;
U_h=1.; % X displacement at x==1.;
%

UV=zeros(2*n,1);
DDL_bloc=[];

for i=1:size(Ind_Node_Front,1)
    if abs(XY_Node(Ind_Node_Front(i),2))<1.e-6
       DDL=2*(Ind_Node_Front(i)-1)+0+1;
       DDL_bloc=[DDL_bloc,DDL];
       UV(DDL)=U_b;
       DDL=2*(Ind_Node_Front(i)-1)+1+1;
       DDL_bloc=[DDL_bloc,DDL];
       UV(DDL)=V_bh;
    end
    
    if abs(XY_Node(Ind_Node_Front(i),2)-1.)<1.e-6
       DDL=2*(Ind_Node_Front(i)-1)+0+1;
       DDL_bloc=[DDL_bloc,DDL];
       UV(DDL)=U_h;
       DDL=2*(Ind_Node_Front(i)-1)+1+1;
       DDL_bloc=[DDL_bloc,DDL];
       UV(DDL)=V_bh;
    end
end 

DDL_bloc=unique(DDL_bloc);

DDL_Tot=1:2*n;
DDL_Libre=setdiff(DDL_Tot,DDL_bloc);

%% solve problem

Vec_B=-K_Mat_meca(:,DDL_bloc)*UV(DDL_bloc);
UV(DDL_Libre)=K_Mat_meca(DDL_Libre,DDL_Libre)\Vec_B(DDL_Libre);
%UV(DDL_Libre)=gmres(K_Mat_meca(DDL_Libre,DDL_Libre),B(DDL_Libre),[],1.e-6,100);

%% post

Strain=B*UV;
Stress=C*Strain;

%% Plot result

% Displacement

h=figure;
set(h,'Name','Displacement');
hold on;
axis equal;
XY_Node_dep=XY_Node+0.2*reshape(UV,2,[])';
plot(XY_Node_dep(:,1),XY_Node_dep(:,2),'r.');
hold off;

h=figure;
set(h,'Name','X Displacement ');
hold on;
axis equal;
plot_cel_vc(Cel,UV(1:2:size(UV,1)));
hold off;

h=figure;
set(h,'Name','Y Displacement');
hold on;
axis equal;
plot_cel_vc(Cel,UV(2:2:size(UV,1)));
hold off;

% Stress

h=figure;
set(h,'Name','XX Stress');
hold on;
axis equal;
plot_cel_vc(Cel,Stress(1:3:size(Stress,1)));
hold off;

h=figure;
set(h,'Name','YY Stress');
hold on;
axis equal;
plot_cel_vc(Cel,Stress(2:3:size(Stress,1)));
hold off;

h=figure;
set(h,'Name','XY Stress');
hold on;
axis equal;
plot_cel_vc(Cel,Stress(3:3:size(Stress,1)));
hold off;

%%














